src/Grids/orthogonal_spherical_shell_grid.jl

```diff
@@ -568,10 +569,63 @@ function OrthogonalSphericalShellGrid(architecture::AbstractArchitecture = CPU()
                      Δzᵃᵃᶜ, Δzᵃᵃᶠ, Azᶜᶜᵃ, Azᶠᶜᵃ, Azᶜᶠᵃ, Azᶠᶠᵃ)
     metric_arrays = map(a -> arch_array(architecture, a), metric_arrays)
 
-    return OrthogonalSphericalShellGrid{TX, TY, TZ}(architecture, Nξ, Nη, Nz, Hx, Hy, Hz, ξ..., η...,
+    grid = OrthogonalSphericalShellGrid{TX, TY, TZ}(architecture, Nξ, Nη, Nz, Hx, Hy, Hz, ξ..., η...,
                                                     coordinate_arrays...,
                                                     metric_arrays...,
                                                     radius)
+
+    fill_metric_halos!(grid)
+
+    return grid
+end
+
+function fill_metric_halos!(grid)
+    Nx, Ny, Nz = size(grid)
+    Hx, Hy, Hz = halo_size(grid)
+
+    metric_arrays = (grid.Δxᶜᶜᵃ, grid.Δxᶠᶜᵃ, grid.Δxᶜᶠᵃ, grid.Δxᶠᶠᵃ,
+                     grid.Δyᶜᶜᵃ, grid.Δyᶜᶠᵃ, grid.Δyᶠᶜᵃ, grid.Δyᶠᶠᵃ,
+                     grid.Azᶜᶜᵃ, grid.Azᶠᶜᵃ, grid.Azᶜᶠᵃ, grid.Azᶠᶠᵃ)
+
+    for metric in metric_arrays
+
+        @inbounds begin
+            nx, ny = Base.size(metric)
+
+            for j in ny-Hy:-1:Hy+1
+                for i in Hx:-1:1
+                    parent(metric)[i, j] = parent(metric)[i+1, j]
```

Comment:
I don't understand exactly what you mean here @simone-silvestri

Reply:
what is the rule for the metrics in the halo? I would expect them to reflect across the boundary following this rule:
```
metric[1-i, j] = metric[i, j]
```
but here you seem to be implying that all metrics in the boundary are equal to the last metric in the interior
```
parent(metric)[i, j] = parent(metric)[i+1, j]
```
what is the correct rule here? I think we need halo metrics to be filled with a fill halo to be sure because neither of the options is always correct.

Reply:
I don’t understand what you mean by “reflect”. I think that in general ambiguous how to fill the metrics and thus I chose to fill them according to their nearest interior value. For FullyConnected topologies, of course, the halo regions correspond to the metrics of the interior neighboring region.

Reply:
It should be like that also for `Periodic` which takes from the other side of the face. In that case (and FullyConnected) the metrics are not so ambiguous. The only instance where metrics in the halos can be considered equal to the interior ones (or ambiguous) is for solid walls and value/gradient BC (where we only need 1 halo). If we don't fill them correctly we might have distortions across the boundary which can crash the simulation/kick off numerical instabilities that grow and pollute the solution (see #3091 where metrics where not correctly considered in reconstructions)

Reply:
Hm.. Could we zoom perhaps?

Reply:
Seems that in lat-lon grid the metrics are the same as the nearest interior point also... no?

```Julia
julia> grid = LatitudeLongitudeGrid(topology=(Bounded, Bounded, Bounded), size = (5, 5, 2), longitude = [0, 1, 3, 6, 12, 20], latitude = [10, 12, 20, 40, 100, 200], z = (-1, 0))
5×5×2 LatitudeLongitudeGrid{Float64, Bounded, Bounded, Bounded} on CPU with 3×3×3 halo and with precomputed metrics
├── longitude: Bounded  λ ∈ [0.0, 20.0]   variably spaced with min(Δλ)=1.0, max(Δλ)=8.0
├── latitude:  Bounded  φ ∈ [10.0, 200.0] variably spaced with min(Δφ)=2.0, max(Δφ)=100.0
└── z:         Bounded  z ∈ [-1.0, 0.0]   regularly spaced with Δz=0.5

julia> grid.Azᶜᶜᵃ
11×11 OffsetArray(::Matrix{Float64}, -2:8, -2:8) with eltype Float64 with indices -2:8×-2:8:
 0.0  2.45431e10  2.44229e10  2.42731e10  9.50055e10  …  -9.39955e11  -3.71217e11  1.06888e12  0.000865158
 0.0  2.45431e10  2.44229e10  2.42731e10  9.50055e10     -9.39955e11  -3.71217e11  1.06888e12  0.000865158
 0.0  2.45431e10  2.44229e10  2.42731e10  9.50055e10     -9.39955e11  -3.71217e11  1.06888e12  0.000865158
 0.0  2.45431e10  2.44229e10  2.42731e10  9.50055e10     -9.39955e11  -3.71217e11  1.06888e12  0.000865158
 0.0  4.90861e10  4.88459e10  4.85461e10  1.90011e11     -1.87991e12  -7.42435e11  2.13776e12  0.00173032
 0.0  7.36292e10  7.32688e10  7.28192e10  2.85016e11  …  -2.81987e12  -1.11365e12  3.20663e12  0.00259547
 0.0  1.47258e11  1.46538e11  1.45638e11  5.70033e11     -5.63973e12  -2.2273e12   6.41327e12  0.00519095
 0.0  1.96344e11  1.95383e11  1.94184e11  7.60044e11     -7.51964e12  -2.96974e12  8.55102e12  0.00692127
 0.0  1.96344e11  1.95383e11  1.94184e11  7.60044e11     -7.51964e12  -2.96974e12  8.55102e12  0.00692127
 0.0  1.96344e11  1.95383e11  1.94184e11  7.60044e11     -7.51964e12  -2.96974e12  8.55102e12  0.00692127
 0.0  1.96344e11  1.95383e11  1.94184e11  7.60044e11  …  -7.51964e12  -2.96974e12  8.55102e12  0.00692127

julia> grid.Azᶜᶜᵃ[1:grid.Nx, 1:grid.Ny]
5×5 Matrix{Float64}:
 2.42731e10  9.50055e10  2.13071e11  2.42295e11  -9.39955e11
 4.85461e10  1.90011e11  4.26141e11  4.8459e11   -1.87991e12
 7.28192e10  2.85016e11  6.39212e11  7.26885e11  -2.81987e12
 1.45638e11  5.70033e11  1.27842e12  1.45377e12  -5.63973e12
 1.94184e11  7.60044e11  1.70456e12  1.93836e12  -7.51964e12
```

Reply:
Btw, there are negative values for `Azᶜᶜᵃ`?

Reply:
> I don’t understand what you mean by “reflect”.

Reflection is a rule like

```julia
for i = 0:-1:H
    a[i] = a[-i + 1]
end
```

This represents a reflection around the interface between i=0 and i=1.

> Does the metric not reflect across the boundaries?

We discussed _continuing_ the metrics across the boundary rather than reflecting them. However, this could have consequences for how fluxes are estimated. Perhaps reflection will make it easier to ensure zero flux for biharmonic closures?

> In general it is ambiguous how one can fill the metrics and thus I chose to fill them according to their nearest interior value

I think it's better to use the real indices not the parent indices. I always think about things in terms of real indices. Also, I don't see the advantage of using `i+1` rather than `i = Hx`. I think this would be better:

```julia
       for j in ny-Hy:-1:Hy+1
                for i in 0:Hx:-Hx+1
                   metric[i, j] = metric[1, j]
```

Reply:
I thought we discussed "continuing the metrics across the boundary" for Bounded topologies and treating them properly for Periodic topologies.

When topology is periodic I don't reflect, I just copy the interior of the other end of the domain.

Look at https://github.com/CliMA/Oceananigans.jl/pull/3239#issuecomment-1711319187

Reply:
OK, "reflect" is a bad choice of word!

Reply:
I'll change it to normal indices -- that's fair point.

Reply:
> I thought we discussed "continuing the metrics across the boundary" for Bounded topologies and treating them properly for Periodic topologies.

Yes, we did, that's what I'm saying! But it's worth clarifying with @simone-silvestri whether he agrees with this strategy (since he's proposing something else)

Reply:
This utility does the "continuation": https://github.com/CliMA/Oceananigans.jl/blob/754708cbf9a9b92a360bfdc9ee6d86972627858e/src/Grids/grid_generation.jl#L17C5-L17C5

Can we generalize that utility to be used with the orthogonal spherical shell grid?

PS I found that code by looking into the rectlinear grid constructor here:

https://github.com/CliMA/Oceananigans.jl/blob/754708cbf9a9b92a360bfdc9ee6d86972627858e/src/Grids/rectilinear_grid.jl#L271

Reply:
Yeah, I found that also and I tried to replicate that functionality. I think I did it correctly.

Reply:
I couldn't use that `lower_exterior_Δcoordᶠ` method; unfortunately I don't remember now why... it was work done over a long flight which hints that I may not have come to that conclusion at my full brain capacity. :( 

I'll revisit and see if I can use that utility here.

Reply:
I wouldn't stress too much about this at the moment. It is kind of a low priority because, in every practical application of the OSSG, the metrics should be filled with a `fill_halo_regions!` (not only for communicating but also for `Periodic` boundary conditions. This might be akin to "extending" in some particular cases but not all). `Bounded` halo metrics matter only for `Value` and `Gradient` boundary conditions

Reply:
Aren't we planning to replace lat-lon grid with OSSG? What happened to that plan?

Reply:
I personally would like to move forward with a fully function OSSG so I can use it for regional simulations (eg in the arctic) that are not restricted to borders that exactly align with geographic coordinates. It will also be nice to rotate domains to follow coastlines, etc

Reply:
Yes we are! #3205
Why did you understood that we are not?